official/rnn_ptb/run_training.py

```diff
@@ -0,0 +1,244 @@
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ==============================================================================
+"""Train a RNN model on the PTB dataset."""
+
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+
+import argparse
+import os
+import shutil
+
+import numpy as np
+import tensorflow as tf
+
+import model
+import model_params
+import util
+
+# Create argument parser. Add the help option only when running this file.
+parser = argparse.ArgumentParser(add_help=(__name__ == '__main__'))
+parser.add_argument(
+    '--data_dir',
+    type=str,
+    default='',
+    help='The directory where train, validation, and test data are saved.')
+parser.add_argument(
+    '--model',
+    type=str,
+    default='medium',
+    help='The model configuration. Options: small, medium(default), large')
+parser.add_argument(
+    '--model_dir',
+    type=str,
+    default='/tmp/ptb_model',
+    help='The directory where the model will be saved.')
+parser.add_argument(
+    '--reset_training',
+    action='store_true',
+    help='Add this option to clear the model directory prior to training.')
+FLAGS, unparsed = parser.parse_known_args()
+
+# Set the train, validation, and test file paths as constants.
+TRAIN_FILE = os.path.join(FLAGS.data_dir, 'ptb.train.txt')
+VALID_FILE = os.path.join(FLAGS.data_dir, 'ptb.valid.txt')
+TEST_FILE = os.path.join(FLAGS.data_dir, 'ptb.test.txt')
+
+# Hard coded number of words in the training file (used to calculate the maximum
+# number steps to train the model).
+TRAIN_FILE_WORDS = 929589
+
+
+def input_fn(input_file, vocab_dict, num_epochs=1):
+  """Generates batches of data from the input file."""
+  def _input_fn(params):
+    batch_size, unrolled_count = params.batch_size, params.unrolled_count
+
+    def batch_data(data, steps_per_epoch):
+      """Batch data so that adjacent batches sequentially traverse the data.
+
+      (See the README for more details)
+
+      Args:
+        data: A sequential array of words with length divisible by batch_size,
+            steps_per_epoch, and unrolled_count.
+        steps_per_epoch: The number of steps it take to complete an epoch.
+      """
+      data = data.reshape([batch_size, steps_per_epoch, unrolled_count])
+      return np.swapaxes(data, 0, 1)
+
+    def input_generator():
+      """Yield batches of the input data."""
+      # Read all words from the input file, and append <eos> at every new line.
+      with tf.gfile.GFile(input_file, 'r') as f:
+        data = f.read().replace('\n', '<eos>').split()
```

Comment:
Do we anticipate users trying files/input with non-unix line endings? Might be worth at least extracting these to module-level vars (something like `INPUT_EOL`, `EOL_TOKEN` or what-have-you), just to make it easier for eventual users to edit. (I speak from the experience of having tried to adapt a similar model in the past...)

Edit: I see below that EOS is defined in `util`; should that be used here as well?

Reply:
Good point, I added a constant for `\n`, and reuse util.EOS for `<eos>`.